providers/siw/siw.c

```diff
@@ -0,0 +1,941 @@
+	wc->wc_flags = 0;
+	wc->imm_data = 0;
+
+	wc->vendor_err = 0;
+	wc->opcode = map_cqe_opcode[cqe->opcode].base;
+	wc->status = map_cqe_status[cqe->status].base;
+	wc->qp_num = (uint32_t)cqe->qp_id;
+}
+
+static int siw_poll_cq(struct ibv_cq *ibcq, int num_entries, struct ibv_wc *wc)
+{
+	struct siw_cq *cq = cq_base2siw(ibcq);
+	int new = 0;
+
+	pthread_spin_lock(&cq->lock);
+
+	for (; num_entries--; wc++) {
+		struct siw_cqe *cqe = &cq->queue[cq->cq_get % cq->num_cqe];
+		atomic_uchar *fp = (atomic_uchar *)&cqe->flags;
+
+		if (_load_mmapped(fp) & SIW_WQE_VALID) {
+			copy_cqe(cqe, wc);
+			_store_mmapped(fp, 0);
+			cq->cq_get++;
+			new++;
+		} else
+			break;
+	}
+	pthread_spin_unlock(&cq->lock);
+
+	return new;
+}
+
+static struct verbs_context_ops siw_context_ops = {
+	.alloc_pd = siw_alloc_pd,
+	.async_event = siw_async_event,
+	.create_ah = siw_create_ah,
+	.create_cq = siw_create_cq,
+	.create_qp = siw_create_qp,
+	.create_srq = siw_create_srq,
+	.dealloc_pd = siw_free_pd,
+	.dereg_mr = siw_dereg_mr,
+	.destroy_ah = siw_destroy_ah,
+	.destroy_cq = siw_destroy_cq,
+	.destroy_qp = siw_destroy_qp,
+	.destroy_srq = siw_destroy_srq,
+	.modify_qp = siw_modify_qp,
+	.modify_srq = siw_modify_srq,
+	.poll_cq = siw_poll_cq,
+	.post_recv = siw_post_recv,
+	.post_send = siw_post_send,
+	.post_srq_recv = siw_post_srq_recv,
+	.query_device = siw_query_device,
+	.query_port = siw_query_port,
+	.query_qp = siw_query_qp,
+	.reg_mr = siw_reg_mr,
+	.req_notify_cq = siw_notify_cq,
+	.resize_cq = siw_resize_cq,
+};
+
+static struct verbs_context *siw_alloc_context(struct ibv_device *base_dev,
+					       int fd, void *pdata)
+{
+	struct siw_context *ctx;
+	struct ibv_get_context cmd;
+	struct siw_alloc_ucontext_resp resp;
+
+	memset(&cmd, 0, sizeof(cmd));
+	memset(&resp, 0, sizeof(resp));
+
+	ctx = verbs_init_and_alloc_context(base_dev, fd, ctx, base_ctx,
+					   RDMA_DRIVER_SIW);
+	if (!ctx)
+		return NULL;
+
+	if (ibv_cmd_get_context(&ctx->base_ctx, &cmd, sizeof(cmd), &resp.base,
+				sizeof(resp))) {
+		verbs_uninit_context(&ctx->base_ctx);
+		free(ctx);
+
+		return NULL;
+	}
+	verbs_set_ops(&ctx->base_ctx, &siw_context_ops);
+	ctx->dev_id = resp.siw.dev_id;
+
+	return &ctx->base_ctx;
+}
+
+static void siw_free_context(struct ibv_context *ibv_ctx)
+{
+	struct siw_context *ctx = ctx_ibv2siw(ibv_ctx);
+
+	verbs_uninit_context(&ctx->base_ctx);
+	free(ctx);
+}
+
+static struct verbs_device *siw_device_alloc(struct verbs_sysfs_dev *unused)
+{
+	struct siw_device *dev;
+
+	dev = calloc(1, sizeof(*dev));
+	if (!dev)
+		return NULL;
+
+	return &dev->base_dev;
+}
+
+static void siw_device_free(struct verbs_device *vdev)
+{
+	struct siw_device *dev =
+		container_of(vdev, struct siw_device, base_dev);
+	free(dev);
+}
+
+static const struct verbs_match_ent rnic_table[] = {
+	VERBS_NAME_MATCH("siw", NULL),
+	{},
+};
+
+static struct verbs_device_ops siw_dev_ops = {
+	.name = "siw",
+	.match_min_abi_version = 0,
+	.match_max_abi_version = INT_MAX,
```

Comment:
No, use the right ABI #, the drivers that use 0, INT_MAX are broken.

Reply:
OK will fix it to one version of 0

Reply:
Probably use 1

Reply:
also the kernel needs to return 1 in the uverbs_abi

Reply:
so let me take version 1 then. I'll have to update the kernel driver then as well, yes.